### ISSUE TITLE
File names of @serializable classes cause error building multiplatform project.

### ISSUE THREAD

**Issue by chrisestesaa01** [Kotlin/kotlinx.serialization#472]
Task 'compileKotlinIos' fails with :
`e: java.util.NoSuchElementException: Sequence contains no element matching the predicate.
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.serializableSyntheticConstructor(GeneratorHelpers.kt:620)
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.serializableSyntheticConstructor(SerializableIrGenerator.kt:30)
        at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.generateSuperSerializableCall(SerializableIrGenerator.kt:130)
...`
with serializable abstract and inherited classes, but only if the name of the file containing the abstract class comes after the inheriting class alphabetically.

**To Reproduce**

In file Cabstract.kt:
`@Serializable
abstract class Cabstract(
    var foo: String
) {
}`

in file Binherit.kt:
`@Serializable
class Binherit(
    var bar: String = ""
) : Cabstract("") {
}`

'./gradlew clean build' results in error. Renaming the file 'Binherit.kt' to 'Dinherit.kt' allows the build to complete with no issue.

**Expected behavior**

Build should run fine regardless of file names.

**Environment**
 - Kotlin version: 1.3.31
 - Library version: 0.11.0
 - Kotlin platforms: Native
 - Gradle version: 5.3.1
 - IDE version Android studio 3.3

**Comment by antanas-arvasevicius** [comment]
Hi,
having the same issue with
kotlin: 1.3.31,  serialization: 0.11.0,  kotlin: native (iOS)

```
e: java.util.NoSuchElementException: Sequence contains no element matching the predicate.
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.IrBuilderExtension$DefaultImpls.serializableSyntheticConstructor(GeneratorHelpers.kt:620)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.serializableSyntheticConstructor(SerializableIrGenerator.kt:30)
	at org.jetbrains.kotlinx.serialization.compiler.backend.ir.SerializableIrGenerator.generateSuperSerializableCall(SerializableIrGenerator.kt:130)
```

still investigating what causing this

**Comment by sandwwraith** [comment]
This looks like more of a separate/cross-module compilation issue. It is definitely not related to the class names.

**Comment by sandwwraith** [comment]
Are your classes split to `main`/`test` modules? Or they're in the same module, but different files? Or in the same file?

**Comment by antanas-arvasevicius** [comment]
We have common module ios module and android module with separate source, ios module depends on common and implements some exact/actual declarations. 
Tests are on the same module with different sourcesets. When running task iosFramework causes this error, dont know if it touches test files

**Comment by sandwwraith** [comment]
So both base class and derived are just in common module?

**Comment by antanas-arvasevicius** [comment]
Yes everything is inside common module

**Comment by sandwwraith** [comment]
Ok, that's fun. It is actually related to names; the problem is the order in which compiler process classes. If file's name with a derived class lexicographically smaller than the file's with a base class, it is processed first and does not see special constructor added by the compiler plugin to the base class. The workaround is to rename files (you don't need to rename classes themselves) or put the whole hierarchy in one file.

**Comment by antanas-arvasevicius** [comment]
So compiler is pretty dumb and don't track usages, but just processes everything lexicographically?

**Comment by antanas-arvasevicius** [comment]
Reminds me #ifndef #define #endif  of C...

**Comment by sandwwraith** [comment]
@antanas-arvasevicius This is more of a serialization plugin issue since it uses separate pipeline after main compilation

**Comment by antanas-arvasevicius** [comment]
Is this can be easily fixed? Should we workaround with adding underscores to file names or wait a fix? What's ETA?

**Comment by sandwwraith** [comment]
@antanas-arvasevicius This will be included in `1.3.40`

**Comment by antanas-arvasevicius** [comment]
Thanks!
